source/extensions/filters/http/cache/cache_filter.cc

```diff
@@ -109,7 +122,13 @@ Http::FilterHeadersStatus CacheFilter::encodeHeaders(Http::ResponseHeaderMap& he
     // Add metadata associated with the cached response. Right now this is only response_time;
     const ResponseMetadata metadata = {time_source_.systemTime()};
     insert_->insertHeaders(headers, metadata, end_stream);
+    if (end_stream) {
+      insert_status_ = CacheInsertStatus::InsertSucceeded;
+    }
```

Comment:
Not necessarily something to address in this change because it's probably out of scope, but there's likely a problem here - if the cache insert operation has to do something asynchronous, the insert has not in fact succeeded at this moment, and worse still, finishing the connection (with `FilterHeadersStatus::Continue`) is going to destroy all the context, which, absent some very careful wrangling on the cache implementation's part, is also going to risk aborting the insert before it completes.

Reply:
You're right that this is something that needs to be addressed. At the moment, there is no way for HttpCache to communicate success or failure from asynchronous requests back to the CacheFilter. This is definitely not ideal.